homeassistant/components/ecovacs/sensor.py

```diff
@@ -183,6 +204,32 @@ async def async_setup_entry(
 
     async_add_entities(entities)
 
+    @callback
+    def _add_legacy_entities(event: CoreEvent) -> None:
+        async_add_entities(
+            EcovacsLegacyLifespanSensor(device, description)
+            for device in controller.legacy_devices
+            for description in LEGACY_LIFESPAN_SENSORS
+            if description.component in device.components
+            and not controller.legacy_entity_is_added(device, description.component)
+        )
+        for device in controller.legacy_devices:
+            for description in LEGACY_LIFESPAN_SENSORS:
+                if description.component in device.components:
+                    controller.add_legacy_entity(device, description.component)
+
+    hass.bus.async_listen("ecovacs_legacy_lifespan_event", _add_legacy_entities)
+
+    def _fire_ecovacs_legacy_lifespan_event(_: Any) -> None:
+        hass.bus.fire("ecovacs_legacy_lifespan_event")
+
+    for device in controller.legacy_devices:
+        config_entry.async_on_unload(
+            device.lifespanEvents.subscribe(
+                _fire_ecovacs_legacy_lifespan_event
+            ).unsubscribe
+        )
```

Comment:
```suggestion
    def _add_legacy_entities(_: Any) -> None:
        entities = []
        for device in controller.legacy_devices:
            for description in LEGACY_LIFESPAN_SENSORS:
                if description.component in device.components and not controller.legacy_entity_is_added(device, description.component):
                    controller.add_legacy_entity(device, description.component)
                    entities.append(EcovacsLegacyLifespanSensor(device, description))
        
        if entities:
            async_add_entities(entities)

    for device in controller.legacy_devices:
        config_entry.async_on_unload(
            device.lifespanEvents.subscribe(
                _add_legacy_entities
            ).unsubscribe
        )
```

Can we no simplify it into one callback function?
If not, can we create a task for the other function instead of firing a event

Your code was looping twice over all device and lifespan sensors. I refactored the suggest to loop only once

Reply:
> Your code was looping twice over all device and lifespan sensors. I refactored the suggest to loop only once

I think it was to late yesterday :see_no_evil: your code looks much cleaner :+1: 

> If not, can we create a task for the other function instead of firing a event

Unfortunately, the `sucks` lib is not async, so even the internal event system does only take sync methods as callbacks for any event subscription. Based on this i didn't get it managed to call the `_add_legacy_entities` from any callback called by the sucks event system - it always end up in a stack trace complaining "no event loop found" :slightly_frowning_face: 
Any hints on how to solve this anyways would be highly appreciated :slightly_smiling_face:

Reply:
@edenhaus do you think we can go with it as is and rework it later?
I would like to have it in 2024.8 :point_right: :point_left:  :see_no_evil:

Reply:
Please try the following:

```diff
diff --git a/homeassistant/components/ecovacs/sensor.py b/homeassistant/components/ecovacs/sensor.py
index 272627af00..4f5be841fd 100644
--- a/homeassistant/components/ecovacs/sensor.py
+++ b/homeassistant/components/ecovacs/sensor.py
@@ -204,8 +204,7 @@ async def async_setup_entry(
 
     async_add_entities(entities)
 
-    @callback
-    def _add_legacy_entities(_: Any) -> None:
+    async def _add_legacy_entities() -> None:
         entities = []
         for device in controller.legacy_devices:
             for description in LEGACY_LIFESPAN_SENSORS:
@@ -221,15 +220,8 @@ async def async_setup_entry(
         if entities:
             async_add_entities(entities)
 
-    for device in controller.legacy_devices:
-        config_entry.async_on_unload(
-            device.lifespanEvents.subscribe(_add_legacy_entities).unsubscribe
-        )
-
-    hass.bus.async_listen("ecovacs_legacy_lifespan_event", _add_legacy_entities)
-
     def _fire_ecovacs_legacy_lifespan_event(_: Any) -> None:
-        hass.bus.fire("ecovacs_legacy_lifespan_event")
+        hass.create_task(_add_legacy_entities())
 
     for device in controller.legacy_devices:
         config_entry.async_on_unload(
```
Tests are successfully on your PR

Reply:
great thanks - that works perfect :+1: